File: journeytomoon/prog.rb
#!/bin/ruby
n, p = gets.split.map(&:to_i)

@nodes = {}
@graph_sizes  = []

def graph_size(size, node, nodes)
  node[:edges].each do |edge_id|
    next if !!nodes[edge_id][:graph_id]
    nodes[edge_id][:graph_id] = node[:graph_id]
    size = graph_size(size+1, nodes[edge_id], nodes)
  end
  size
end
  
p.times do
   a,b = gets.split.map(&:to_i)
   
   @nodes[a] ||= { id: a, edges: [] }
   @nodes[a][:edges] << b

   @nodes[b] ||= { id: b, edges: [] }
   @nodes[b][:edges] << a
end

@nodes.each do |id, node|
  next if !!node[:graph_id]
  node[:graph_id] = id
  @graph_sizes << graph_size(1, node, @nodes)
end

single_nodes = (n - @nodes.length)
@count = 0
@graph_sizes.length.times do |i|
  graph_size = @graph_sizes.shift
  remaining_nodes = @graph_sizes.inject(:+).to_i + single_nodes
  @count += graph_size * remaining_nodes
end

@count += (single_nodes - 1).downto(1).inject(:+) if single_nodes > 1

puts @count

File: journeytomoon/prog.swift
import Foundation;
var line = readLine()!.components(separatedBy: [" "])
let n = Int(line[0])!
let p = Int(line[1])!

class Node {
    var id : Int!
    var edges: [Int]
    var graph_id: Int?

    init(id: Int) {
        self.id = id
        self.edges = []
    }
}

var nodes = [Int: Node]()
var graph_sizes = [Int]()

func graph_size(size: inout Int, node: Node, nodes: inout [Int: Node]) -> Int {
  for edge_id in node.edges {
    if(nodes[edge_id]?.graph_id != nil){ continue }
    nodes[edge_id]?.graph_id = node.graph_id
    size+=1
    size = graph_size(size: &size, node: nodes[edge_id]!, nodes: &nodes)
  }
  return size
}

for _ in 0..<p {
  var line = readLine()!.components(separatedBy: [" "])
  var a = Int(line[0])!
  var b = Int(line[1])!
  nodes[a] = nodes[a] ?? Node(id: a)
  nodes[a]?.edges.append(b)
  nodes[b] = nodes[b] ?? Node(id: b)
  nodes[b]?.edges.append(a)
}

for (id,node) in nodes {
  if(node.graph_id != nil) { continue }
  node.graph_id = id
  var size = 1
  var gs = graph_size(size: &size, node: node, nodes: &nodes)
  graph_sizes.append(gs)
}

var single_nodes = (n - nodes.count)
var count = 0
for _ in 0..<graph_sizes.count {
  var gs = graph_sizes.removeFirst()
  var remaining_nodes = graph_sizes.reduce(0,+) + single_nodes
  count += gs * remaining_nodes
}

if(single_nodes > 1) { for i in 1..<single_nodes { count += i }}

print("\(count)")
